Add the InfluxDB Timestream connector

trevorbonas opened the pull request:
*Issue #, if available:*

N/A.

*Description of changes:*

The InfluxDB Timestream connector has been added.

The InfluxDB Timestream connector is a Rust application that receives, translates, and ingests line protocol data into Timestream for LiveAnalytics. The connector is intended to be deployed as part of a CloudFormation stack as a Lambda function, but can also be used as a library or run locally as a Lambda function.

The connector includes the following components:
- A README with documentation for how the connector translates line protocol data to Timestream records, a guide for using the connector, and documentation for different configuration options.
- A SAM template that deploys a REST API Gateway, using either synchronous or asynchronous deployment, and the connector as a Lambda function.

A few things to note:

The connector splits up batches and ingests records in chunks of 100 at a time in parallel. This means that if a record in one of these chunks has an issue, an error will occur and ingestion will stop but other records in the request batch, records that were processed before or in parallel with the chunk with the erroneous record, will be successfully ingested. When asynchronous invocation is used, the entire request will be added to the Lambda's dead letter queue. Rejected records and AWS SDK for Rust errors are logged. This will help users narrow down the problem among the line protocol points in their batch.

The connector uses the InfluxDB v3 parser, since InfluxDB v3 is written in Rust and its parser is available as a crate. This parser has a few inconsistencies in behavior compared to the InfluxDB v2 parser.

103 tests have been created for the connector, 40 integration tests and 63 unit tests. The following is a list of all tests with a short summary for each:
- [`test_mtmm_basic`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L110-L144) integration test: Tests ingesting a single line protocol point.
- [`test_mtmm_create_database`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L147-L197) integration test: Tests ingesting a single line protocol point and creating a database.
- [`test_mtmm_unusual_query_parameters`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L200-L234) integration test: Tests ingesting a single line protocol point with a query parameters keys with unusual spelling.
- [`test_mtmm_no_query_parameters`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L237-L267) integration test: Tests ingesting a single line protocol point without query parameters.
- [`test_mtmm_multiple_timestamps`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L270-L305) integration test: Tests ingesting a single line protocol point with two timestamps. This test expects failure.
- [`test_mtmm_many_tags_many_fields`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L308-L354) integration test: Test ingesting a single line protocol point with 50 tags and 50 fields.
- [`test_mtmm_float`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L357-L391) integration test: Tests ingesting a single line protocol point with a float value for the field.
- [`test_mtmm_string`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L394-L428) integration test: Tests ingesting a single line protocol point with a string value for the field.
- [`test_mtmm_bool`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L431-L465) integration test: Tests ingesting a single line protocol point with a bool value for the field.
- [`test_mtmm_max_tag_length`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L468-L506) integration test: Tests ingesting a single line protocol point with a tag where the tag's key has length 60, the maximum allowed dimension name length, and its value is 1988 characters long. The length of the tag key and tag value together amount to the maximum size for a dimension pair, 2 kilobytes.
- [`test_mtmm_beyond_max_tag_length`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L509-L545) integration test: Tests ingesting a single line protocol point with a tag where the tag's key has length 60, the maximum allowed dimension name length, and its value is 1989 characters long. The length of the tag key and tag value together exceed the maximum size for a dimension pair, 2 kilobytes. This test expects failure.
- [`test_mtmm_max_field_length`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L548-L585) integration test: Tests ingesting a single line protocol point with a field where the length of the field key is the maximum measure name length, 256 chars, and the length of the field value is the maximum measure value size, 2048.
- [`test_mtmm_beyond_max_field_length`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L588-L623) integration test: Tests ingesting a single line protocol point with a field where the length of the field key is the maximum measure name length, 256, and the length of the field value is beyond the maximum measure value size, 2048. This test expects failure.
- [`test_mtmm_max_unique_field_keys`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L626-L667) integration test: Tests ingesting a batch of line protocol points where the number of unique field keys in the batch equals the maximum number of allowed unique measures for a single table, 1024.
- [`test_mtmm_beyond_max_unique_field_keys`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L670-L710) integration test: Tests ingesting a batch of line protocol points where the number of unique field keys in the batch exceeds the maximum number of allowed unique measures for a single table, 1024. This test expects failure.
- [`test_mtmm_max_unique_tag_keys`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L713-L754) integration test: Tests ingesting a batch of line protocol points where the number of unique tag keys in the batch equals the maximum number of allowed unique dimensions for a single table, 128.
- [`test_mtmm_beyond_max_unique_tag_keys`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L757-L797) integration test: Tests ingesting a batch of line protocol points where the number of unique tag keys in the batch exceeds the maximum number of unique dimensions for a single table, 128. This test expects failure.
- [`test_mtmm_max_table_name_length`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L800-L836) integration test: Tests ingesting a single line protocol point with a measurement with length equal to the maximum number of bytes a table name can have.
- [`test_mtmm_beyond_max_table_name_length`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L839-L873) integration test: Tests ingesting a single line protocol point with a measurement with length exceeding the maximum number of bytes a table name can have. This test expects failure.
- [`test_mtmm_nanosecond_precision`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L876-L912) integration test: Tests ingesting a single line protocol point with nanosecond precision.
- [`test_mtmm_microsecond_precision`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L915-L949) integration test: Tests ingesting a single line protocol point with microsecond precision.
- [`test_mtmm_second_precision`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L952-L986) integration test: Tests ingesting a single line protocol point with second precision. 
- [`test_mtmm_no_precision`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L989-L1026) integration test: Tests ingesting a single line protocol point without precision specified.
- [`test_mtmm_empty_point`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1029-L1051): Tests ingesting an empty string.
- [`test_mtmm_small_timestamp`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1054-L1088) integration test: Tests ingesting a single line protocol point with a single-digit millisecond timestamp.
- [`test_mtmm_5_measurements`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1091-L1133) integration test: Tests ingesting a batch of line protocol points with five measurements.
- [`test_mtmm_100_measurements`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1136-L1178) integration test: Tests ingesting a batch of line protocol points with one-hundred measurements.
- [`test_mtmm_5000_batch`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1181-L1221) integration test: Tests ingesting a batch of 5000 line protocol points with a single measurement.
- [`test_mtmm_no_credentials`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1225-L1260) integration test: Tests ingesting without AWS credentials. This test expects a panic.
- [`test_mtmm_incorrect_credentials`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1264-L1305) integration test: Tests ingesting with incorrect AWS credentials. This tests expects a panic.
- [`test_mtmm_custom_dimension_partition_key_optional_enforcement`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1308-L1346) integration test: Tests ingesting a single line protocol point and specifying a valid configuration for a custom dimension partition key with optional enforcement. 
- [`test_mtmm_custom_dimension_partition_key_required_enforcement_accepted`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1349-L1388) integration test: Tests ingesting a single line protocol point and specifying a valid configuration for a custom dimension partition key with required enforcement and a successful ingestion.
- [`test_mtmm_custom_dimension_partition_key_required_enforcement_rejected`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1391-L1429) integration test: Tests ingesting a single line protocol point and specifying a valid configuration for a custom dimension partition key with required enforcement and an unsuccessful ingestion. This test expects failure.
- [`test_mtmm_custom_dimension_partition_key_no_dimension`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1432-L1469) integration test: Tests ingesting a single line protocol point and specifying a configuration for a custom dimension partition key without a dimension specified. This test expects failure.
- [`test_mtmm_custom_dimension_partition_key_no_enforcement`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1472-L1509) integration test: Tests ingesting a single line protocol point and specifying a configuration for a custom dimension partition key without an enforcement configuration specified. This test expects failure.
- [`test_mtmm_custom_measure_partition_key`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1512-L1550) integration test: Tests ingesting a single line protocol point and specifying a valid configuration for a custom measure partition key.
- [`test_mtmm_custom_measure_partition_key_with_dimension`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1553-L1591) integration test: Tests ingesting a single line protocol point and specifying a valid configuration for a custom measure partition key with a dimension specified. The dimension should be ignored.
- [`test_mtmm_custom_measure_partition_key_with_enforcement`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1594-L1633) integration test: Tests ingesting a single line protocol point and specifying a valid configuration for a custom measure partition key with an enforcement configuration specified. The enforcement configuration should be ignored.
- [`test_stmm_basic`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1636-L1670) integration test: Tests ingesting a single line protocol point using the single-table multi measure schema.
- [`test_stmm_varying_metrics`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs#L1673-L1730) integration test: Tests ingesting a batch with 100 different measurements using single-table multi measure schema.
- [`test_get_precision_query_string_parameters_array`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L243-L249) unit test: Tests getting the precision from an incoming request where queryStringParameters is an array, with the precision contained within the array.
- [`test_get_precision_query_string_parameters_object`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L252-L258) unit test: Tests getting the precision from an incoming request where queryStringParameters is a JSON object, with the precision contained within the object with the key "precision".
- [`test_get_precision_query_string_parameters_object_nanoseconds`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L261-L267) unit test: Tests getting the precision from an incoming request where queryStringParameters is a JSON object and the precision is nanoseconds.
- [`test_get_precision_query_string_parameters_object_microseconds`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L270-L276) unit test: Tests getting the precision from an incoming request where queryStringParameters is a JSON object and the precision is microseconds.
- [`test_get_precision_query_string_parameters_object_seconds`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L279-L285) unit test: Tests getting the precision from an incoming request where queryStringParameters is a JSON object and the precision is seconds.
- [`test_get_precision_query_parameters_array`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L288-L294) unit test: Tests getting the precision from an incoming request where queryStringParameters is instead named queryParameters and is a JSON object and the precision is provided as an array.
- [`test_get_precision_query_parameters_object`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L297-L303) unit test: Tests getting the precision from an incoming request where queryStringParameters is instead named queryParameters and is a JSON object and the precision is provided as a JSON object.
- [`test_get_precision_incorrect_query_parameters_key`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L306-L310) unit test: Tests getting the precision from an incoming request where the queryStringParameters key is neither queryStringParameters nor queryParameters. This test expects failure.
- [`test_get_precision_incorrect_precision_key`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L313-L317) unit test: Tests getting the precision from an incoming request where the precision key has been named incorrectly.
- [`test_parse_field_integer`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L62-L77) unit test: Tests parsing a line protocol point with an integer field value.
- [`test_parse_field_float`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L80-L95) unit test: Tests parsing a line protocol point with a float field value.
- [`test_parse_field_string_double_quote`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L98-L116) unit test: Tests parsing a line protocol point with a string field value using double quotes.
- [`test_parse_field_string_single_quote`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L119-L137) unit test: Tests parsing a line protocol point with a string field value using single quotes.
- [`test_parse_field_boolean`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L140-L155) unit test: Tests parsing a line protocol point with a boolean field value.
- [`test_parse_field_boolean_invalid`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L158-L166): Tests parsing a line protocol point with an invalid boolean field value. This test expects failure.
- [`test_parse_measurement_unescaped_equals`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L169-L187) unit test: Tests parsing a line protocol point where the measurement name includes an unescaped equals sign.
- [`test_parse_measurement_underscore_begin`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L190-L207) unit test: Tests parsing a line protocol point where the measurement name begins with an underscore.
- [`test_parse_no_fields`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L210-L218) unit test: Tests parsing a line protocol point without any fields.
- [`test_parse_multiple_fields`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L221-L239) unit test: Tests parsing a line protocol point with multiple fields.
- [`test_parse_multiple_measurements`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L243-L253) unit test: Test parsing a line protocol point with multiple measurement names. This test expects failure. This test is currently skipped due to an inconsistency between the InfluxDB v2 and v3 line protocol parsers. An issue relating to this inconsistency has been opened [here](https://github.com/influxdata/influxdb3_core/issues/29).
- [`test_parse_no_timestamp`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L256-L264) unit test: Tests parsing a line protocol point without a timestamp. This test expects failure.
- [`test_parse_non_unix_timestamp`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L267-L276) unit test: Tests parsing a line protocol point with a non-unix timestamp. This test expects failure.
- [`test_parse_timestamp_with_quotes`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L279-L286) unit test: Tests parsing a line protocol point where the timestamp is in double quotations. This test expects failure.
- [`test_parse_no_whitespace`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L289-L297) unit test: Tests parsing a line protocol point with no whitespace between components. This test expects failure.
- [`test_parse_multiple_timestamps`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L300-L307) unit test: Tests parsing a line protocol point with multiple timestamps. This test expects failure.
- [`test_parse_batch`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L310-L334) unit test: Tests parsing multiple line protocol points with integer field values.
- [`test_parse_emojis`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L337-L356) unit test: Tests parsing a line protocol point with emojis.
- [`test_parse_escaped_comma`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L359-L375) unit test: Tests parsing a line protocol point with escaped commas included in different parts of the point.
- [`test_parse_escaped_equals`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L378-L394) unit test: Tests parsing a line protocol point with escaped equals signs in different parts of the point.
- [`test_parse_unescaped_equals_measurement`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L397-L412) unit test: Tests parsing a line protocol point with an unescaped equals sign in the measurement name.
- [`test_parse_escaped_equals_measurement`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L415-L430) unit test: Tests parsing a single line protocol point with an unescaped equals sign in the measurement name.
- [`test_parse_escaped_space`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L433-L449) unit test: Tests parsing a line protocol point with escaped spaces in different parts of the point.
- [`test_parse_measurement_escaped_space`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L452-L467) unit test: Tests parsing a line protocol point with an escaped space in the measurement name.
- [`test_parse_escaped_double_quote_field_value`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L470-L488) unit test: Tests parsing a line protocol point with escaped quotes in the field value.
- [`test_parse_non_escaped_double_quote_field_value`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L491-L499) unit test: Tests parsing a line protocol point with unescaped quotes in the field value. This test expects failure.
- [`test_parse_escaped_backslash_field_value`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L502-L520) unit test: Tests parsing a line protocol point with escaped backslashes in the field value.
- [`test_parse_escaped_backslash_not_field_value`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L523-L539) unit test: Tests parsing a line protocol point with escaped backslashes in different parts of the point, excluding the field value.
- [`test_parse_single_point_single_comment`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L542-L563) unit test: Tests parsing a line protocol point with one comment included.
- [`test_parse_single_point_multiple_comments`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L566-L588) unit test: Tests parsing a line protocol point with two comments included.
- [`test_parse_multiple_points_single_comment`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L591-L615) unit test: Tests parsing two line protocol points with one comment included.
- [`test_parse_multiple_points_multiple_comments`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L618-L643) unit test: Tests parsing two line protocol points with two comments included.
- [`test_parse_nanoseconds_timestamp`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L646-L667) unit test: Tests parsing a line protocol point with a nanosecond timestamp.
- [`test_parse_microseconds_timestamp`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L670-L690) unit test: Tests parsing a line protocol point with a microsecond timestamp.
- [`test_parse_milliseconds_timestamp`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L693-L713) unit test: Tests parsing a line protocol point with a millisecond timestamp.
- [`test_parse_seconds_timestamp`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L716-L736) unit test: Tests parsing a line protocol point with a second timestamp.
- [`test_parse_empty`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser/tests.rs#L739-L748) unit test: Tests parsing empty line protocol.
- [`test_mtmm_single_record`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/records_builder/tests.rs#L8-L64) unit test: Tests building a single record with multi-table multi measure schema.
- [`test_mtmm_single_destination`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/records_builder/tests.rs#L67-L153) unit test: Tests building records where all records go to the same table with multi-table multi measure schema.
- [`test_mtmm_multi_record`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/records_builder/tests.rs#L156-L243) unit test: Tests building records where records to go different tables with multi-table multi measure schema.
- [`test_mtmm_empty_dimensions`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/records_builder/tests.rs#L246-L293) unit test: Tests building records with empty dimensions for multi-table multi measure schema.
- [`test_mtmm_varying_timestamp_records`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/records_builder/tests.rs#L296-L394) unit test: Tests building records where records have varying timestamps for multi-table multi measure schema.
- [`test_stmm_single_record`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/records_builder/tests.rs#L397-L449) unit test: Tests building a single record with one measure for single-table multi measure schema.
- [`test_stmm_multi_record`](https://github.com/awslabs/amazon-timestream-tools/blob/4c518202dbec86c55759af96a723890c02a1ac83/integrations/influxdb_connector/influxdb-timestream-connector/src/records_builder/tests.rs#L452-L532) unit test: Tests building records with different measurement names destined for the same table with single-table multi measure schema.
- [`test_parse_tags_from_str_empty_string`](https://github.com/awslabs/amazon-timestream-tools/blob/e56ddeee2b6947e4b111bb2c2edd6149450ab78f/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L338-L343) unit test: Tests parsing tags where the entire tag is an empty string.
- [`test_parse_tags_from_str_whitespace_only`](https://github.com/awslabs/amazon-timestream-tools/blob/e56ddeee2b6947e4b111bb2c2edd6149450ab78f/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L346-L351) unit test: Tests parsing tags where the entire tag is whitespace.
- [`test_parse_tags_from_str_tag_with_no_value`](https://github.com/awslabs/amazon-timestream-tools/blob/e56ddeee2b6947e4b111bb2c2edd6149450ab78f/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L354C8-L361) unit test: Tests parsing tags where the tag has a key and no value.
- [`test_parse_tags_from_str_multiple_tags`](https://github.com/awslabs/amazon-timestream-tools/blob/e56ddeee2b6947e4b111bb2c2edd6149450ab78f/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L364C8-L375) unit test: Tests parsing multiple tags separated by commas followed by a space.
- [`test_parse_tags_from_str_extra_commas`](https://github.com/awslabs/amazon-timestream-tools/blob/e56ddeee2b6947e4b111bb2c2edd6149450ab78f/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L378C8-L387) unit test: Tests parsing tags where an extra comma has been inserted between two tags.
- [`test_parse_tags_from_str_multiple_equals`](https://github.com/awslabs/amazon-timestream-tools/blob/e56ddeee2b6947e4b111bb2c2edd6149450ab78f/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L390C8-L397) unit test: Tests parsing tags where the key is followed by multiple values separated by equals signs.
- [`test_parse_tags_from_str_trimming_whitespace`](https://github.com/awslabs/amazon-timestream-tools/blob/e56ddeee2b6947e4b111bb2c2edd6149450ab78f/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L400C8-L409) unit test: Tests parsing tags where whitespace is included in various parts of the tag.
- [`test_parse_tags_from_str_tag_with_empty_value_explicit`](https://github.com/awslabs/amazon-timestream-tools/blob/e56ddeee2b6947e4b111bb2c2edd6149450ab78f/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L412-L423) unit test: Tests parsing tags where some of the tags are missing a value.
- [`test_parse_tags_from_str_error_empty_key`](https://github.com/awslabs/amazon-timestream-tools/blob/e56ddeee2b6947e4b111bb2c2edd6149450ab78f/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L426-L430) unit test: Tests parsing tags missing a key. This test expects failure.
- [`test_parse_tags_from_str_error_only_equals`](https://github.com/awslabs/amazon-timestream-tools/blob/e56ddeee2b6947e4b111bb2c2edd6149450ab78f/integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs#L433-L437) unit test: Tests parsing tags where only an equals sign is included, without a key nor a value. This test expects failure.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

forestmvey committed:
Initial commit for the InfluxDB Timestream Connector.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

trevorbonas committed:
Add pre-commit hook for secrets scanning

*Issue #, if available:*

N/A.

*Description of changes:*

- A pre-commit hook has been added that uses `aws-secrets` in order to prevent secrets from being committed.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

trevorbonas committed:
Add SAM template

*Issue #, if available:*

N/A

*Description of changes:*

- SAM template `template.yml` added.
    - Asynchronous and synchronous invocation supported.
- Documentation for how the connector can be deployed using the SAM template added.
- `lambda_runtime` crate added.
- `LambdaEvent<serde_json::Value>` used instead of `lambda_http::Request`, in order to support more types of requests, instead of just AWS service integrations.
- Tests added for handling different kinds of `queryParameters` keys in requests.
- DLQ implemented for asynchronous invocation.
- Integration tests changed to check the newly returned `serde_json::Value` struct.

- [x] Unit tests passed.
- [x] Integration tests passed.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

trevorbonas committed:
Add support for customer-defined partition keys

- Users can now define partition keys for their newly-created tables. Partition key configuration is controlled using three environment variables, `custom_partition_key_type`, `custom_partition_key_dimension`, and `enforce_custom_partition_key`.
- Environment variables added for CDPK support.
- SAM template parameters added for CDPK support.
- Documentation added for CDPK support.
- Integration tests added for CDPK support.
- Integration test asserts moved to end of tests, in order to ensure resources are cleaned up.

- [x] Unit tests passed.
- [x] Integration tests passed.

trevorbonas committed:
Improve IAM permissions

*Issue #, if available:*

N/A.

*Description of changes:*

- Deployment permissions have been updated according to the required permissions as discovered by testing deploying the connector using its SAM template.
- `samconfig.toml` added with default stack deployment options.
- "Troubleshooting" section added to the README with two known errors users may encounter.
    - Issue with cross-platform Rust compilation.
    - ConflictExceptions occurring with concurrent instances of the connector.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

trevorbonas committed:
Return 2.0 formatted Lambda responses for local invocation

- The environment variable `local_invocation` has been added and is set to `true` by default when the Lambda function is run with `cargo lambda watch`. This environment enables responses to be returned in a format `cargo lambda` expects, in the 2.0 formatting. Otherwise, the 1.0 format will be returned, which the synchronous API Gateway expects.
- `cargo fmt` run.

- [x] Tested locally.
- [x] Tested with synchronous invocation.
- [x] Tested with asynchronous invocation.

trevorbonas committed:
Optimize connector

*Issue #, if available:*

N/A.

*Description of changes:*
- Ingestion to multiple tables done in parallel.
- Ingestion of 100 records to a single table done in parallel.
- Chunking of records into batches of 100 done in parallel.
- Limit on maximum possible number of threads added.
- Print statement for each 100 records removed.
- Logging option added to SAM template.
- All `println!` calls changed to `info!`.
- Default logging level for the connector changed to `INFO`.
- Trace statements that measure the execution time of each function added.
- Instructions added to README for how to configure logging levels.
- Database and tables are no longer checked if their corresponding environment variables for creation are not set.
- The checking of whether tables exist has been moved within the asynchronous code block used to ingest records to a table. This checking is now done in parallel and the hashmap is looped through once, instead of twice.
- Instructions added to README for how to reduce stack costs.

- [x] Integration tests passed.
- [x] Unit tests passed.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

forestmvey committed:
Add Single Table Mapping for InfluxDB Timestream Connector (#23)

* Initial implementation for adding single-table mapping.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

* Adding environment variables and updating documentation.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

* Adding tests and revising single-table mapping.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

* Adding single table example to README.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

* Fixing measure-name using line protocol metric name for single table mapping.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

* Fix typo in README.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

* Fixing documentation typos and test setting wrong environment variable.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

* Fixing table for single table multi measure records example in README.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

* Fix invalid line protocol examples with additional comma separating the timestamp.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

---------

Signed-off-by: forestmvey <forestv@bitquilltech.com>

forestmvey committed:
Multi-table Mapping Default for InfluxDB Timestream Connector (#25)

* Set default table mapping to multi-table for the InfluxDB Timestream Connector.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

* Fixing table formatting in README for InfluxDB Timestream Connector.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

* Updating README to reference multi-table for the default table mapping.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

---------

Signed-off-by: forestmvey <forestv@bitquilltech.com>

trevorbonas committed:
Add support for returning gzip compressed responses (#29)

* Add gzip support to Go client

* Add gzip support to template

* Ignore custom_partition_key_type if it is invalid option

* Add comment about lack of local gzip support

forestmvey committed:
Defining Least-privilege Lambda Invocation Permissions for the InfluxDB Timestream Connector (#33)

* Defining least-privilege Lambda invocation permissions for the connector. Adding output for IAM policy when deploying using the SAM template.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

* Making stage name dynamic for output least privilege IAM policy.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

* Adding section in README for ingestion permissions of Lambda function.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

* Revising wording for IAM permissions in README.

Signed-off-by: forestmvey <forestv@bitquilltech.com>

---------

Signed-off-by: forestmvey <forestv@bitquilltech.com>

fredjoonpark committed:
Adding support for KMS key and tags in `create_database` (#35) (#36)

Added support for supplying CMK and tags in `create_database` and `create_table`

pdhamoth commented on integrations/influxdb_connector/influxdb-timestream-connector/README.md:
```diff
@@ -0,0 +1,670 @@
+                "apigateway:PATCH"
+            ],
+            "Resource": [
+                "arn:aws:apigateway:{region}::/restapis/*",
+                "arn:aws:apigateway:{region}::/restapis",
+                "arn:aws:apigateway:{region}::/tags/*"
```
Can we make this scoped to the resource name pattern, similar to the other policies

trevorbonas replied:
Unfortunately not. REST API ARNs use the following format: `arn:partition:apigateway:region::/restapis/api-id`. [Link (under `RestApi`)](https://docs.aws.amazon.com/apigateway/latest/developerguide/arn-format-reference.html#apigateway-v1-arns). The `api-id` is random and is not related to the REST API gateway's name pattern.

pdhamoth commented on integrations/influxdb_connector/influxdb-timestream-connector/README.md:
```diff
@@ -0,0 +1,670 @@
+- The REST API Gateway incurs the highest costs for the stack. Reduce REST API Gateway costs by including as many line protocol points in a request as possible. 5,000-20,000 line protocol points in each request is ideal. The Lambda memory size, Lambda timeout, REST API Gateway timeout, and client timeout may have to be increased in order to accommodate larger requests.
+    - NOTE: if requests contain only a single line protocol point, the ratio of REST API Gateway requests to ingested records would be 1:1 and costs would be much higher than including multiple line protocol points in each request.
+- If possible, create the necessary tables in advance, to reduce the time the connector will take to create tables. The connector adds a one second delay for table or database creation in order to avoid throttling.
+- Reduce the amount of memory the connector uses as a Lambda function. The default, and smallest possible value, is 128 MB.
```
We have recommended to batch up to 20k points in each request, can we also add ideal lambda memory config on such workloads

trevorbonas replied:
Are 20k points recommended for ingesting to Timestream for InfluxDB or to Timestream for LiveAnalytics? If it is the recommended batch size for Timestream for InfluxDB, is this batch size still relevant when ingesting to Timestream for LiveAnalytics?

Timestream for LiveAnalytics can only handle 100 records in a batch. This connector will split up batches and ingest a 100 at a time to Timestream for LiveAnalytics in parallel.

pdhamoth replied:
Just want to make it a bit clear, neither referring to influx or LA batch for 20k points. Referring to this statement in line number 652, regarding the batching for this connector API and lambda handler

> 5,000-20,000 line protocol points in each request is ideal.

comment, is to get clarity on what should be the ideal lambda memory for a stable workload to this API and can we recommend

trevorbonas replied:
I see. I will updated the sample application to use 20k points and make a note of the recommended amount of memory for that batch size.

trevorbonas replied:
I have lowered the recommended maximum supported points based on testing. I have added a note suggesting an increase in Lambda memory and troubleshooting ingesting large batches.

pdhamoth commented on integrations/influxdb_connector/influxdb-timestream-connector/docs/adr/require-line-protocol-tags.md:
````diff
@@ -0,0 +1,27 @@
+```
+readings,fleet=Alberta velocity=52.13 1725050416
+```
+All components of a line protocol point are required except for tags.
````
nit: tags are required ?

trevorbonas replied:
Line protocol [does not require tags](https://docs.influxdata.com/influxdb/v2/reference/syntax/line-protocol/#tag-set).

The InfluxDB Timestream connector, however, requires tags in line protocol. This was a design decision explained in this document.

pdhamoth commented on integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs:
```diff
@@ -0,0 +1,438 @@
+use std::sync::Arc;
+use std::time::Instant;
+use std::{str, thread, time};
+use timestream_utils::*;
```
Can we avoid glob import `*`  ? use alias or import only specific methods as needed

trevorbonas replied:
I have removed all glob imports and now only import necessary specific methods.

pdhamoth commented on integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs:
```diff
@@ -0,0 +1,438 @@
+) -> Result<(), Error> {
+    // Handle parsing body in request
+
+    let line_protocol = str::from_utf8(body).unwrap();
```
nit: can we gracefully handle errors with logs and not just rely on panic to help customer identify the error if it's related to the payload format

trevorbonas replied:
I have added error handling for this. Users will get an error log and a returned error.

pdhamoth commented on integrations/influxdb_connector/influxdb-timestream-connector/src/line_protocol_parser.rs:
```diff
@@ -0,0 +1,81 @@
+        for (tag_key, tag_value) in tag_set {
+            new_tags.push((tag_key.to_string(), tag_value.to_string()));
+        }
+    }
```
We parse through all the lines in the payload. Can we maintain some state to track datapoints with same tags, possibly we could use common attributes on write batches with same tags in the downstream  and save write cost.
Also, for `multi-table-multi-measure`, if measure name will be constant, can we add that to common attribute as well ?

trevorbonas replied:
All records are now ingested with common attributes. The common attributes are measure name, dimensions, timestamp precision, and measure value type.

pdhamoth commented on integrations/influxdb_connector/influxdb-timestream-connector/src/records_builder.rs:
```diff
@@ -0,0 +1,180 @@
+    }
+    if std::env::var("database_name").is_err() {
+        return Err(anyhow!("database_name environment variable is not defined"));
+    }
```
we are loading these static env variables multiple times. Can we organize all the env variable loading to a util and use `OnceCell` and just load all of them once and use them across

trevorbonas replied:
`OnceCell` is now used for environment variables. `lib.rs` and `timestream_utils.rs` have their own environment variable configuration, to support adding more parsers and ingestors in the future that don't need the Timestream environment variables to be set.

pdhamoth commented on integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs:
```diff
@@ -0,0 +1,438 @@
+            if let Ok(true) = std::env::var("enable_table_creation").map(env_var_to_bool) {
+                let _ =
+                    create_table_if_non_existent(&client_clone, &database_name_clone, &table_name)
+                        .await;
```
Can we optimize to maintain a local table cache once a table is created to avoid network calls as we could potentially create 1000s of tables for `multi_table_multi_measures`

trevorbonas replied:
A `HashSet` is now used to track previously created tables. The `retry_with_backoff` function also checks for conflict exceptions, so that different invocations don't erroneously try to create existing tables.

pdhamoth commented on integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs:
```diff
@@ -0,0 +1,438 @@
+            }
+            Ok(response)
+        }
+        // An Err is required in order to send messages to the Lambda's
+        // dead letter queue, when the connector is deployed as part of a stack
+        // with asynchronous invocation
+        Err(error) => Err(anyhow!(error.to_string())),
+    }
```
How do we help differentiate between a `retryable` v/s `non-retryable` ?
Also, can we add clarity on the format of DLQ messages between different error types (`CRUD / Write` , `retryable v/s non-retryables`) and how they can be deserialized from the queue and attempted for retries

trevorbonas replied:
All messages for errors now include what operation was being attempted (such as record ingestion or table creation) and whether the error is retryable. For AWS SDK for Rust errors, I added the function `is_retryable_status_code` to check the status code of errors.

pdhamoth commented on integrations/influxdb_connector/influxdb-timestream-connector/tests/integration_test.rs:
```diff
@@ -0,0 +1,1732 @@
+                Ok(_) => (),
+
+                Err(error) => {
+                    println!(
+                        "Table deletion failed for table {}: {:?}",
+                        table_name_to_delete,
+                        error.raw_response()
```
nit: can we use error! ?

trevorbonas replied:
These calls and other prints in helper functions can be changed to `error!` or `info!`. From my understanding, the actual integration tests don't support logging without initializing the logger at the start of every test.

trevorbonas replied:
I have confirmed that logs for this entire file does not work. To show any output with `cargo test -- --nocapture` `println!` must be used instead, as it currently is.

pdhamoth commented on integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs:
```diff
@@ -0,0 +1,438 @@
+    match table_exists(client, database_name, table_name).await {
+        Ok(true) => (),
+        Ok(false) => {
+            thread::sleep(time::Duration::from_secs(TIMESTREAM_API_WAIT_SECONDS));
```
we still have concurrency of `16` threads and could potentially hit throttling errors leading to wasted work for the failed tables as default CRUD TPS is `1`
Can we add some jitters with backoff to retry

trevorbonas replied:
I have added the function `retry_with_backoff` which uses jitters. This function can be used with any kind of request.

pdhamoth commented on integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs:
```diff
@@ -0,0 +1,438 @@
+        // Create a future for ingesting to the current table
+        let future = task::spawn(async move {
+            if let Ok(true) = std::env::var("enable_table_creation").map(env_var_to_bool) {
+                let _ =
+                    create_table_if_non_existent(&client_clone, &database_name_clone, &table_name)
+                        .await;
```
nit: can we add logs for table and database creation failures

trevorbonas replied:
Added.

pdhamoth commented on integrations/influxdb_connector/influxdb-timestream-connector/src/lib.rs:
```diff
@@ -0,0 +1,438 @@
+        Err(_) => None,
+    };
+
+
```
nit: extra line

pdhamoth commented on integrations/influxdb_connector/influxdb-timestream-connector/src/timestream_utils.rs:
```diff
@@ -0,0 +1,396 @@
+    {
+        Ok(_) => {}
+        Err(error) => {
+            info!("SdkError: {:?}", error.raw_response().unwrap());
```
nit: error!

trevorbonas replied:
Changed.

pdhamoth reviewed (changes requested):
Thanks for adding comprehensive test cases

trevorbonas committed:
Remove wildcard imports and redundant map

trevorbonas committed:
Handle errors when decoding body to UTF-8

trevorbonas committed:
Address InfluxDB Timestream connector feedback (#40)

This PR addresses the feedback given on the [upstream PR](https://github.com/awslabs/amazon-timestream-tools/pull/201) to add the InfluxDB Timestream connector.

Changes:
- Use Rust doc comments.
- Move `parse_tags_from_str` into `timestream_utils.rs`.
- Change an `info` log to `error` for failed record ingestion.
- Add retries with exponential backoff and jitter, usable for any kind of request.
- Use `OnceCell` for environment variables.
- Keep a cache of created table names to reduce API calls.
- Use common attributes for record ingestion.
- Identify all errors as retryable or non-retryable.
- Increase returned SDK error verbosity.
- Add `once-cell` crate for thread-safe `OnceCell`.
- Add `async-trait` crate to support `OnceCell` usage.
- Add `tokio-test` crate for async doc tests.
- Document the `EnableAsyncInvocation` SAM template parameter.
- Add caveats for duplicate timestamps and troubleshooting timeouts.

- [X] Doc tests passed.
- [X] Integration tests passed.
- [X] Unit tests passed.

pdhamoth reviewed (approved):
lgtm. thanks
